### PR TITLE
feat(bentocloud): implement deployment CLI

### PR DESCRIPTION
**Pull request by Haivilo**
## What does this PR address?
It implements the CLI interface for bentocloud deployment.

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_internal/cloud/client.py [hunk 1]
```diff
@@ -49,7 +49,11 @@ def __init__(self, endpoint: str, api_token: str) -> None:
 
     def _is_not_found(self, resp: requests.Response) -> bool:
         # Forgive me, I don't know how to map the error returned by gorm to juju/errors
-        return resp.status_code == 400 and "record not found" in resp.text
+        return (
+            resp.status_code == 404
+            or resp.status_code == 400
```
Comment (aarnphm) [inline src/bentoml/_internal/cloud/client.py:54]: Why are we return of it is 404?


Ah I was always meant to submit a PR for error improvement on this end. Eh maybe when we refactor cloud_rest_client

File: src/bentoml/_internal/cloud/client.py [hunk 2]
```diff
@@ -414,9 +418,9 @@ def get_models_list(self) -> ModelWithRepositoryListSchema | None:
         self._check_resp(resp)
         return schema_from_json(resp.text, ModelWithRepositoryListSchema)
 
-    def get_deployment_list(self, cluster_name: str) -> DeploymentListSchema | None:
+    def get_deployment_list(self, cluster_name: str, params: dict[str,str|int] | None = None) -> DeploymentListSchema | None:
```
Comment (aarnphm) [inline src/bentoml/_internal/cloud/client.py:421]: make to run `pre-commit run --all`

File: src/bentoml/_internal/cloud/deployment.py [hunk 3]
```diff
@@ -45,13 +46,7 @@ def for_hpa_conf(cls, **kwargs) -> DeploymentTargetHPAConf:
 
     @classmethod
     def for_runner(cls, **kwargs) -> DeploymentTargetRunnerConfig:
-        exclusive_api_server_key = {
-            v for v in kwargs if v not in attr.fields_dict(DeploymentTargetRunnerConfig)
-        }
-        return bentoml_cattr.structure(
-            {k: v for k, v in kwargs.items() if k not in exclusive_api_server_key},
-            DeploymentTargetRunnerConfig,
-        )
```
Comment (aarnphm) [inline src/bentoml/_internal/cloud/deployment.py:54]: Why are we removing this?

File: src/bentoml/_internal/cloud/deployment.py [hunk 4]
```diff
@@ -149,7 +144,15 @@ def update(
...
-    ):
+    ) -> DeploymentSchema:
+        if mode is None:
+            mode = DeploymentMode.Function
+        if type is None:
+            type = DeploymentTargetType.STABLE
+        if kube_namespace is None:
+            kube_namespace = default_kube_namespace
+        if cluster_name is None:
+            cluster_name = default_cluster_name
```
Comment (aarnphm) [inline src/bentoml/_internal/cloud/deployment.py:155]: use first_not_none(value, default=default)

File: src/bentoml_cli/deployment.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+        if "," in v:
+            parsed += tuple(v.split(","))
+        else:
+            parsed += tuple(v.split())
+        parsedLs.append(
+            dict(map(lambda x: x.strip()[1:-1].split("="), filter(lambda x: x, parsed)))
+        )
+        # Remove trailing space then single quotation mark in front and end, split by = to make a key,val pair
+        # --runner a b c d
+        # ("'a'", "'b'")
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:88]:
```suggestion
```

File: src/bentoml_cli/deployment.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    return parsedLs
+
+
+def add_deployment_command(cli: click.Group) -> None:
+    from bentoml_cli.utils import BentoMLCommandGroup
+    from bentoml._internal.configuration.containers import BentoMLContainer
+
+    client = BentoMLContainer.bentocloud_client.get()
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:96]:
```suggestion
    client = BentoMLContainer.bentocloud_client.get()
   
   	output = click.option("-o", '--output', help='display the output of this command', type=click.Choice(['json', 'porcelain']))
```

Comment (aarnphm) [reply src/bentoml_cli/deployment.py:96]: Then you can use this decorator for all of the command to process it.

File: src/bentoml/_internal/cloud/client.py (hunk 1 above)
Comment (Haivilo) [reply src/bentoml/_internal/cloud/client.py:54]: I found that now when the model or deployment doesn't exist, it returns 404.

Comment (frostming) [reply src/bentoml/_internal/cloud/client.py:54]: This is changed by me, we have fixed the error handling of BentoCloud that correctly returns 404 for not found.

File: src/bentoml_cli/deployment.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        "--runner",
+        cls=NargsOptions,
+        multiple=True,
+        callback=parse_runners_callback,
+        default=None,
+        help="Runners.",
+    )
+    def create(  # type: ignore
+        deployment_name: str,
+        bento: str,
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:177]:
```suggestion
        bento: str,
        output: t.Literal['default', 'json']
```

File: src/bentoml/_internal/cloud/deployment.py (hunk 3 above)
Comment (aarnphm) [reply src/bentoml/_internal/cloud/deployment.py:54]: I guess we set forbid_extra_keys to False
we should just set forbid_extra_keys to true

File: src/bentoml_cli/deployment.py [hunk 8]
```diff
@@ -83,41 +85,67 @@ def parse_runners_callback(
         # ("'a'", "'b'")
     return parsedLs
 
+
 class Mutex(click.Option):
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:89]: It is not really Mutex option anymore is it? I think this is RequiredAllMutexOption

File: src/bentoml_cli/deployment.py [hunk 9]
```diff
@@ -83,41 +85,67 @@ def parse_runners_callback(
...
+        options = [click.argument(
+        "deployment_name",
+        type=click.STRING,
+        required=True,
+    ),
+    click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    ),
+    click.option(
+        "--cluster-name", type=click.STRING, default=None, help="Name of the cluster."
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:136]: run formatter please 😄

File: src/bentoml_cli/deployment.py [hunk 10]
```diff
@@ -127,23 +155,35 @@ def deployment_cli():
...
+            "resource_instance",
+            "min_replicas",
+            "max_replicas",
+            "type",
+            "mode",
+            "runner",
+            "force",
+            "threads",
+            "push"
+        ],
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:178]: I still think having to implement this as a eager feature works the best. This is not ideal imho

File: src/bentoml_cli/deployment.py (hunk 8 above)
Comment (Haivilo) [reply src/bentoml_cli/deployment.py:89]: It differs a bit imo, here `[--deployment-name, --bento]` and `[--file]` are required & mutually exclusive, but other options except `--output` are forbidden too when using `--file`. but they are optional

File: src/bentoml_cli/deployment.py (hunk 10 above)
Comment (Haivilo) [reply src/bentoml_cli/deployment.py:178]: The issue with eager is that it is unable to check if `--file` and `--deployment_name` are provided together (we only want one to be provided). I will check if we can add that check somehow with eager.

File: src/bentoml_cli/deployment.py (hunk 8 above)
Comment (aarnphm) [reply src/bentoml_cli/deployment.py:89]: so if users pass in --file it should be executed immediately right?

Comment (aarnphm) [reply src/bentoml_cli/deployment.py:89]: in terms of setting context and the other variable for create from file we can tell user with the --file workflow to use environment variable to set for that variable specifically

File: src/bentoml_cli/deployment.py (hunk 10 above)
Comment (aarnphm) [reply src/bentoml_cli/deployment.py:178]: if --file is passed then it ignore everything else if it is an eager option right? we confirmed that behaviour the other day?

Comment (Haivilo) [reply src/bentoml_cli/deployment.py:178]: Yes that's right, but eager is processed first so it cannot disallow commands such as `bentoml deployment create -f create.json --deployment-name test-1`. However `--deployment-name` here is omitted. Not sure if we want to forbid extra commands when `-f/--file` is present? Currently it'll give a warning of `Error: Illegal usage: 'file' is mutually exclusive with 'deployment_name'` in that situation.

File: src/bentoml_cli/deployment.py (hunk 8 above)
Comment (Haivilo) [reply src/bentoml_cli/deployment.py:89]: `--output` and `--context` are valid options when using `--file`. I agree `context` could be stored as an environment variable, but `--output` might be better to be passed in?...

File: src/bentoml_cli/deployment.py (hunk 10 above)
Comment (aarnphm) [reply src/bentoml_cli/deployment.py]: What I'm saying is in the case of `--file` user need to put into the environment variable for deployment-name and cluster-name

Comment (aarnphm) [reply src/bentoml_cli/deployment.py]:
```bash
BENTOCLOUD_DEPLOYMENT_NAME=... BENTOCLOUD_CLUSTER_NAME=... bentoml deploy -f create.json
```

File: src/bentoml_cli/deployment.py [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+        type: DeploymentTargetType,
+        mode: DeploymentMode,
+        runner: dict[str, t.Any],
+        output: t.Literal["json", "default"],
+        force: bool,
+        threads: int,
+        push: bool,
+    ) -> DeploymentSchema:
+        """Create a deployment."""
+        if file is not None:
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:308]: add a _echo here just sayiing that --file is passed, then only context and cluster_name will be get from environment variable

File: src/bentoml_cli/deployment.py [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+        type: DeploymentTargetType,
+        mode: DeploymentMode,
+        runner: dict[str, t.Any],
+        output: t.Literal["json", "default"],
+    ) -> DeploymentSchema:
+        """Update a deployment"""
+        if file is not None:
+            return client.deployment.update_from_file(
+                path=file, context=context, cluster_name=cluster_name
+            )
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:467]:
```suggestion
            return client.deployment.update_from_file(path_or_stream=file)
```

File: src/bentoml_cli/deployment.py [hunk 13]
```diff
@@ -0,0 +1,637 @@
...
+
+    def shared_decorator(f: t.Callable[..., t.Any]):
+        options = [
+            click.argument(
+                "deployment_name",
+                type=click.STRING,
+                required=True,
+            ),
+            click.option(
+                "--context", type=click.STRING, default=None, help="Yatai context name."
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:150]:
```suggestion
                "--context", type=click.STRING, default=None, help="Yatai context name.", envvar=...
```

File: src/bentoml/_internal/cloud/config.py [hunk 14]
```diff
@@ -18,8 +18,6 @@
 logger = logging.getLogger(__name__)
 
 default_context_name = "default"
-default_cluster_name = "default"
-default_kube_namespace = "yatai"
```
Comment (aarnphm) [inline src/bentoml/_internal/cloud/config.py:22]: why are we removintg this?

File: src/bentoml_cli/deployment.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+        ]
+        for opt in reversed(options):
+            f = opt(f)
+        return f
+
+    @cli.group(name="deployment", cls=BentoMLCommandGroup)
+    def deployment_cli():
+        """Deployment Subcommands Groups"""
+
+    @deployment_cli.command(context_settings={"ignore_unknown_options": True})
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:61]:
```suggestion
    @deployment_cli.command()
```

File: src/bentoml_cli/deployment.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        elif output == "json":
+            click.echo(json.dumps(bentoml_cattr.unstructure(res), indent=2))
+        return res
+
+    @deployment_cli.command()
+    @click.option(
+        "-f",
+        "--file",
+        type=click.File(),
+        required=True,
```
Comment (aarnphm) [inline src/bentoml_cli/deployment.py:91]:
```suggestion
```

**Review by aarnphm — APPROVED** [review]

**Pull request merged**